lib/ansible/module_utils/basic.py

```diff
@@ -1885,28 +1886,37 @@ def do_cleanup_files(self):
         for path in self.cleanup_files:
             self.cleanup(path)
 
-    def exit_json(self, **kwargs):
-        ''' return from the module, without error '''
+    def _return_formatted(self, kwargs):
+
         self.add_path_info(kwargs)
-        if not 'changed' in kwargs:
-            kwargs['changed'] = False
+
         if 'invocation' not in kwargs:
             kwargs['invocation'] = {'module_args': self.params}
+
+        if 'warnings' not in kwargs and self.warnings:
+            kwargs['warnings'] = self.warnings
```

Comment:
Shouldn't this also assert the "shape" of both the `warnings` arg and/or the warnings that were passed in (ie, that it's a list)? Maybe make self.warnings a getter-only property instead so it can't be overwritten accidentally (instead just appended to)?